[FIX] account: update repartition lines correctly

asno-odoo opened the pull request:
When repartition_lines are added to an already existing account_tax record, on upgrade, the new repartition lines are inserted with just tags, which results in the insertion of null value in document_type column.

To reproduce:

- Modify `l10n_fr/__manifest__.py` version to `2.0`
- Create an empty DB
- Start `./odoo-bin -c ../.myodoorc -i l10n_fr` with demo data
- Stop the database (can dump it for convenience)
- Start `./odoo-bin -c ../.myodoorc -u l10n_fr`
- Add a couple of repartition lines (invoice + refund) to an existing tax to `odoo/addons/l10n_fr/data/template/account.tax-fr.csv`
- Modify `l10n_fr/__manifest__.py` version to `2.1`
- Start `./odoo-bin -c ../.myodoorc -u l10n_fr`

```
Traceback (most recent call last):
  File "/home/odoo/src/odoo/saas-16.3/odoo/service/server.py", line 1302, in preload_registries
    registry = Registry.new(dbname, update_module=update_module)
  File "<decorator-gen-16>", line 2, in new
  File "/home/odoo/src/odoo/saas-16.3/odoo/tools/func.py", line 87, in locked
    return func(inst, *args, **kwargs)
  File "/home/odoo/src/odoo/saas-16.3/odoo/modules/registry.py", line 90, in new
    odoo.modules.load_modules(registry, force_demo, status, update_module)
  File "/home/odoo/src/odoo/saas-16.3/odoo/modules/loading.py", line 478, in load_modules
    processed_modules += load_marked_modules(env, graph,
  File "/home/odoo/src/odoo/saas-16.3/odoo/modules/loading.py", line 366, in load_marked_modules
    loaded, processed = load_module_graph(
  File "/home/odoo/src/odoo/saas-16.3/odoo/modules/loading.py", line 232, in load_module_graph
    migrations.migrate_module(package, 'post')
  File "/home/odoo/src/odoo/saas-16.3/odoo/modules/migration.py", line 233, in migrate_module
    migrate(self.cr, installed_version)
  File "/home/odoo/src/odoo/saas-16.3/addons/l10n_fr/migrations/2.1/post-migrate_update_taxes.py", line 8, in migrate
    env['account.chart.template'].try_loading('fr', company)
  File "/home/odoo/src/odoo/saas-16.3/addons/account/models/chart_template.py", line 142, in try_loading
    return self._load(template_code, company, install_demo)
  File "/home/odoo/src/odoo/saas-16.3/addons/account/models/chart_template.py", line 186, in _load
    self._load_data(data)
  File "/home/odoo/src/odoo/saas-16.3/addons/account/models/chart_template.py", line 499, in _load_data
    created_vals[model] = self.with_context(lang='en_US').env[model]._load_records(create_vals)
  File "/home/odoo/src/odoo/saas-16.3/odoo/models.py", line 4663, in _load_records
    data['record']._load_records_write(data['values'])
  File "/home/odoo/src/odoo/saas-16.3/odoo/models.py", line 4594, in _load_records_write
    self.write(values)
  File "/home/odoo/src/odoo/saas-16.3/addons/account/models/account_tax.py", line 362, in write
    return super().write(self._sanitize_vals(vals))
  File "/home/odoo/src/odoo/saas-16.3/odoo/models.py", line 4033, in write
    field.write(self, value)
  File "/home/odoo/src/odoo/saas-16.3/odoo/fields.py", line 4240, in write
    self.write_batch([(records, value)])
  File "/home/odoo/src/odoo/saas-16.3/odoo/fields.py", line 4261, in write_batch
    self.write_real(records_commands_list, create)
  File "/home/odoo/src/odoo/saas-16.3/odoo/fields.py", line 4441, in write_real
    flush()
  File "/home/odoo/src/odoo/saas-16.3/odoo/fields.py", line 4397, in flush
    comodel.create(to_create)
  File "<decorator-gen-119>", line 2, in create
  File "/home/odoo/src/odoo/saas-16.3/odoo/api.py", line 415, in _model_create_multi
    return create(self, arg)
  File "/tmp/tmp90kuz8ug/migrations/util/orm.py", line 210, in wrapper
    return f(*args, **kwargs)
  File "/tmp/tmp90kuz8ug/migrations/base/0.0.0/pre-models-match_uniq.py", line 25, in create
    return super().create(vals_list)
  File "<decorator-gen-12>", line 2, in create
  File "/home/odoo/src/odoo/saas-16.3/odoo/api.py", line 415, in _model_create_multi
    return create(self, arg)
  File "/home/odoo/src/odoo/saas-16.3/odoo/models.py", line 4231, in create
    records = self._create(data_list)
  File "/home/odoo/src/odoo/saas-16.3/odoo/models.py", line 4434, in _create
    cr.execute(
  File "/home/odoo/src/odoo/saas-16.3/odoo/sql_db.py", line 319, in execute
    res = self._obj.execute(query, params)
psycopg2.errors.NotNullViolation: null value in column "document_type" of relation "account_tax_repartition_line" violates not-null constraint
DETAIL:  Failing row contains (305, null, 1, 1, 1, 1, tax, f, 2023-10-17 19:52:56.357571, 2023-10-17 19:52:56.357571, 100, 6, null).

```

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/148643).

lordkrandel commented:
Cannot replicate, this is what I'm doing:

- Comment your change
- Modify `l10n_fr/__manifest__.py` version to 2.0
- Create an empty DB
- Start `./odoo-bin -c ../.myodoorc -i l10n_fr` with demo data
- Add a repartition line on both invoice and credit note with account and tags
- Stop the database
- Modify `l10n_fr/__manifest__.py` version to 2.1
- Start `./odoo-bin -c ../.myodoorc -u l10n_fr`
- The migration script runs... no error?

Am I doing something wrong? Do you have any ticket to look at?
<details><summary>Open me <--</summary>
<p>
    <img src="https://github.com/odoo/odoo/assets/1665365/74e72abd-ced2-4a1f-adb7-882f01275515"/>
    <img src="https://github.com/odoo/odoo/assets/1665365/58e5b78c-a3c3-41ec-aff0-fb5a6f1d16fc"/>
    <img src="https://github.com/odoo/odoo/assets/1665365/b8c56f9c-0279-4cae-a6a9-4f11d95437a1"/>
</p>
</details>

lordkrandel commented:
Nvm, deleted the old comment, I understood and replicated the error.

lordkrandel reviewed (changes requested):
Can you please update the commit message with the extended steps I've added?
Besides that, it **LGTM**.

I also have a question: we create the `[old]` tax only if we change tax_amount or add/remove repartition lines - but if we change the accounts or tags, we just overwrite...?

asno-odoo commented:
> Can you please update the commit message with the extended steps I've added? Besides that, it **LGTM**.
> 
> I also have a question: we create the `[old]` tax only if we change tax_amount or add/remove repartition lines - but if we change the accounts or tags, we just overwrite...?

Thanks for adding steps in the PR. I have added the extra steps to the commit message.
According to [this](https://github.com/odoo/odoo/blob/17.0/addons/account/models/chart_template.py#L296-L305) we just overwrite tags. For account I am not sure.

apan-odoo commented:
Hey, @qdp-odoo can we have your review on this?
I have a db migrating to v17 which is blocking at this point.
[upg-1217501](https://upgrade.odoo.com/web#id=1217501&cids=1&menu_id=107&action=150&model=upgrade.request&view_type=form)
[opw-3668781](https://www.odoo.com/web#id=3668781&cids=2&menu_id=4720&action=333&active_id=70&model=project.task&view_type=form)
cc: @lordkrandel @asno-odoo

qdp-odoo commented on addons/account/models/chart_template.py:
```diff
@@ -255,6 +255,7 @@ def tax_template_changed(tax, template):
             return (
                 tax.amount_type != template.get('amount_type', 'percent')
                 or tax.amount != template.get('amount', 0)
+                or len(tax.repartition_line_ids) != len(values.get('repartition_line_ids'))
```
```suggestion
                or len(tax.repartition_line_ids) != len(template.get('repartition_line_ids'))
```
Let's keep it consistent, it makes it clearer... don't you think?

qdp-odoo commented:
@asno-odoo made a suggestion, but it's already fine TBH. Feel free to ignore it and r+ yourself

@robodoo delegate+

asno-odoo commented:
@robodoo r+
